lib/core/src/receive_swap.rs

```diff
@@ -260,12 +260,87 @@ impl ReceiveSwapStateHandler {
                 err: format!("Cannot transition from {from_state:?} to Complete state"),
             }),
 
-            (Created, TimedOut) => Ok(()),
+            (Created | TimedOut, TimedOut) => Ok(()),
             (_, TimedOut) => Err(PaymentError::Generic {
                 err: format!("Cannot transition from {from_state:?} to TimedOut state"),
             }),
 
+            (Complete, Failed) => Err(PaymentError::Generic {
+                err: format!("Cannot transition from {from_state:?} to Failed state"),
+            }),
             (_, Failed) => Ok(()),
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use std::sync::Arc;
+
+    use anyhow::Result;
+
+    use crate::{
+        model::PaymentState::*,
+        test_utils::{new_persister, new_receive_swap, new_receive_swap_state_handler},
+    };
+
+    #[tokio::test]
+    async fn test_receive_swap_state_transitions() -> Result<()> {
+        let (_temp_dir, storage) = new_persister()?;
+        let storage = Arc::new(storage);
+
+        let receive_swap_state_handler = new_receive_swap_state_handler(storage.clone())?;
+
+        // Test valid combinations of states
+        let valid_combinations = [
+            (Created, Pending),
+            (Pending, Pending),
+            (Created, Complete),
+            (Pending, Complete),
+            (Created, TimedOut),
+            (TimedOut, TimedOut),
+            (Created, Failed),
+            (Pending, Failed),
+            (TimedOut, Failed),
+        ];
+
+        for (first_state, second_state) in valid_combinations.iter() {
+            let receive_swap = new_receive_swap(Some(*first_state));
+            storage.insert_receive_swap(&receive_swap)?;
+
+            assert!(receive_swap_state_handler
+                .update_swap_info(&receive_swap.id, *second_state, None, None)
+                .await
+                .is_ok());
+        }
+
+        // Test invalid combinations of states
+        let invalid_combinations = [
```

Comment:
Why not threat the invalid combinations as `all_possible_combinations - valid_combinations`?

E.g. I'd generate all possible combinations, then iterate:
- if it's in the valid list, perform the test expected for valid ones
- if not, perform the test expected for invalid ones